src/Database/PostgreSQL/PQTypes/Internal/State.hs

```diff
@@ -1,49 +1,273 @@
 -- | Definition of internal DBT state.
 module Database.PostgreSQL.PQTypes.Internal.State
-  ( DBState (..)
+  ( -- * ConnectionData
+    ConnectionData
+  , getConnectionSource
+  , getConnectionAcquisitionModeIO
+  , withConnectionData
+  , changeAcquisitionModeTo
+  , withConnection
+
+    -- * DBState
+  , DBState (..)
+  , mkDBState
   , updateStateWith
   ) where
 
+import Control.Concurrent.MVar.Lifted
+import Control.Monad
+import Control.Monad.Base
+import Control.Monad.Catch
+import Data.Function
+import Data.Typeable
 import Foreign.ForeignPtr
+import GHC.Stack
 
+import Data.Monoid.Utils
 import Database.PostgreSQL.PQTypes.FromRow
+import Database.PostgreSQL.PQTypes.Internal.BackendPid
 import Database.PostgreSQL.PQTypes.Internal.C.Types
 import Database.PostgreSQL.PQTypes.Internal.Connection
+import Database.PostgreSQL.PQTypes.Internal.Exception
 import Database.PostgreSQL.PQTypes.Internal.QueryResult
+import Database.PostgreSQL.PQTypes.SQL
 import Database.PostgreSQL.PQTypes.SQL.Class
 import Database.PostgreSQL.PQTypes.Transaction.Settings
 
+data ConnectionState cdata
+  = OnDemand
+  | Acquired !IsolationLevel !Permissions !Connection !cdata
+  | Finalized
+
+-- Note: initConnectionState and finalizeConnectionState are invoked inside
+-- bracket and run with asynchronous exceptions softly masked. However, both of
+-- them may run queries that start/finish a transaction. Running queries is a
+-- blocking (and thus interruptible) operation, but if these queries are
+-- interrupted with an asynchronous exception, then a connection is leaked, so
+-- they need to be run with asynchronous exceptions hard masked with
+-- uninterruptibleMask.
+
+initConnectionState
+  :: MonadBase IO m
+  => InternalConnectionSource m cdata
+  -> ConnectionAcquisitionMode
+  -> m (ConnectionState cdata)
+initConnectionState ics = \case
+  AcquireOnDemand -> pure OnDemand
+  AcquireAndHold tsIsolationLevel tsPermissions -> do
+    let initSql =
+          smconcat
+            [ "BEGIN"
+            , case tsIsolationLevel of
+                DefaultLevel -> ""
+                ReadCommitted -> "ISOLATION LEVEL READ COMMITTED"
+                RepeatableRead -> "ISOLATION LEVEL REPEATABLE READ"
+                Serializable -> "ISOLATION LEVEL SERIALIZABLE"
+            , case tsPermissions of
+                DefaultPermissions -> ""
+                ReadOnly -> "READ ONLY"
+                ReadWrite -> "READ WRITE"
+            ]
+    (conn, cdata) <- takeConnection ics
+    _ <- liftBase . uninterruptibleMask_ $ runQueryIO @SQL conn initSql
+    pure $ Acquired tsIsolationLevel tsPermissions conn cdata
+
+finalizeConnectionState
+  :: (HasCallStack, MonadBase IO m)
+  => InternalConnectionSource m cdata
+  -> ExitCase r
+  -> ConnectionState cdata
```

Comment:
Nit: I think the only reason the connection source here is present is to be able to instantiate/finalize the connection right? What about embedding that directly in the constructors in `ConnectionState`. So something like

```haskell
data ConnectionState cdata
  = OnDemand (IO Connection) (Connection -> IO ()) -- I think `ContT Connection IO ()` might also work.
  | Acquired !IsolationLevel !Permissions !Connection !cdata (Connection -> IO ())
  | Finalized
```

I don't have a strong opinion here, so I'll leave the choice to you.